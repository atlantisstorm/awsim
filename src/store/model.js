import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const handleFetchData = ({ state, accessKeyId, secretAccessKey, region }) => {
  let instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D
  if (process.env.NODE_ENV === "development") {
    instances = extractInstances(testInstances);
  } else {
    AWS.config.update({
      "accessKeyId": accessKeyId,
      "secretAccessKey": secretAccessKey,
      "region": region
    });

    const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});
    const params = {
      DryRun: false
    };

    ec2.describeInstances(params, function(err, data) {
      if(err) {
        console.log("Error", err.stack);
      } else {
        instances = extractInstances(data);
      }
    });
  }

  return { ...state, loading: false, instances };
}

export const handleFilterInstances = ({ state }) => {
  const filteredInstances = filterInstances(state.instances, state.filters);
  return { ...state, filteredInstances };
}

export const handleFilterUpdate = ({ state, filterName, filterValue }) => {
  const filters = state.filters;
  const updatedFilter = { ...filters[filterName], currentValue: filterValue };
  const updatedFilters = { ...filters, [filterName]: updatedFilter };
  const filteredInstances = filterInstances(state.instances, updatedFilters);

  return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };    
}

export const handleModalSetDisplay = ({ state, display }) => {
  const newModalSettings = {
    ...state.modalSettings,
    display: display
  };

  return { ...state, modalSettings: newModalSettings };
}

export const handleModalShowInstance = ({ state, instanceId, pageX, pageY}) => {
  const newModalSettings = {
    ...state.modalSettings,
    pageX,
    pageY,
    display: true,    
    instance: getInstanceRawData(state.instances, instanceId)
  };

  return { ...state, selectedInstance: instanceId, modalSettings: newModalSettings };
}

export const handleRefreshData = ({ state }) => {
  return { ...state, loading: true };
}

export const handleSortOptionsSet = ({ state, field, order }) => {
  return { ...state, sortOptions: setSort(field, order) };
}

export const Model = {
  handleFetchData,
  handleFilterInstances,
  handleFilterUpdate,
  handleModalSetDisplay,
  handleModalShowInstance,
  handleRefreshData,
  handleSortOptionsSet
};

export default Model;