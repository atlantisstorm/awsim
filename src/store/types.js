// Emmm... action types.
export const FETCH_DATA = "FETCH_DATA";
export const FILTER_INSTANCES = "FILTER_INSTANCES";
export const FILTER_UPDATE = "FILTER_UPDATE";
export const MODAL_SET_DISPLAY = "MODAL_SET_DISPLAY";
export const MODAL_SHOW_INSTANCE = "MODAL_SHOW_INSTANCE";
export const REFRESH_DATA = "REFRESH_DATA";
export const SORT_OPTIONS_SET = "SORT_OPTIONS_SET";