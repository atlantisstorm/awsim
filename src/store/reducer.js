import { useReducer } from 'react';
import config from '../data/config.json';
import testInstances from '../data/running.instances.json';
import { extractInstances, getInstanceRawData, initFilters,  setSort } from '../utils';
import {
  FETCH_DATA,
  FILTER_INSTANCES,
  FILTER_UPDATE,
  MODAL_SET_DISPLAY,
  MODAL_SHOW_INSTANCE,
  REFRESH_DATA,
  SORT_OPTIONS_SET
} from './types';

const AWS  = require('aws-sdk');

export const defaultState = {
  loading: true,
  instances: [],
  selectedInstance: '',
  filteredInstances: [],
  filters: initFilters(),
  filtersOrdered : config.filters,
  fields: config.fields,
  modalSettings: {
    display: false,
    pageX: 5,
    pageY: 5,
    instance: {}
  },
  sortOptions: setSort('Name', 'asc'),
  aws: {
    accessKeyId: config.aws.accessKeyId,
    secretAccessKey: config.aws.secretAccessKey
  }
};

const filterInstances = (instances, filters) => {
  const filteredInstances = instances.filter(instance =>{
    let includeInstance = true;

    for (const key in filters) {
      const filter = filters[key];
      const filterField = filter.field;
      const filterValue = filter.currentValue;
      const filterMatch = filter.match;
      if (filterField === "Regions") {
        // The 'Regions' filter is only applied on loading/refresh request in useFetchData, so ignore here.
        continue;
      }

      if (filterMatch !== "regex") {
        if (filterValue && instance[key] !== filterValue) {
          includeInstance = false;
          break;
        }
      } else {
        // Currently just the "Name" field.
        const filterPattern = new RegExp('^' + filterValue);
        if (!filterPattern.test(instance[filterField])) {
          includeInstance = false;
          break;
        }
      }
    }

    return includeInstance;
  });

  return filteredInstances;
}

export const reducer = (state, action) => {
  let payload;
  let instances;
  let filters;
  let newModalSettings;

  switch (action.type) {
    case FETCH_DATA: {
      payload = action.payload;
      instances = state.instances; // We do this to stop test data being over-written, bit hacky! :D

      if (process.env.NODE_ENV === "development") {
        instances = extractInstances(testInstances);
      } else {
        AWS.config.update({
          "accessKeyId": payload.accessKeyId,
          "secretAccessKey": payload.secretAccessKey,
          "region": payload.region
        });

        const ec2 = new AWS.EC2({apiVersion: config.aws.apiVersion});

        const params = {
          DryRun: false
        };

        ec2.describeInstances(params, function(err, data) {
          if(err) {
            console.log("Error", err.stack);
          } else {
            instances = extractInstances(data);
          }
        });
      }

      return { ...state, loading: false, instances };
    }

    case FILTER_INSTANCES: {
      const filteredInstances = filterInstances(state.instances, state.filters);
      return { ...state, filteredInstances };
    }

    case FILTER_UPDATE: {
      payload = action.payload;
      filters = state.filters;

      const filterName = payload.name;
      const filterValue = payload.value;
      const updatedFilter = { ...filters[filterName], currentValue: filterValue };
      const updatedFilters = { ...filters, [filterName]: updatedFilter };

      const filteredInstances = filterInstances(state.instances, updatedFilters);
      return { ...state, filteredInstances, filters: { ...filters, [filterName]: updatedFilter } };
    }

    case MODAL_SET_DISPLAY: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: payload.display
      };
      return { ...state, modalSettings: newModalSettings };
    }

    case MODAL_SHOW_INSTANCE: {
      payload = action.payload;

      newModalSettings = {
        ...state.modalSettings,
        display: true,
        pageX: payload.pageX, 
        pageY: payload.pageY,
        instance: getInstanceRawData(state.instances, payload.instanceId)
      };
      return { ...state, selectedInstance: payload.instanceId, modalSettings: newModalSettings };
    }

    case REFRESH_DATA: {
      // This will trigger FETCH_DATA
      return { ...state, loading: true };
    }

    case SORT_OPTIONS_SET: {
      // This will trigger FETCH_DATA
      payload = action.payload;
      return { ...state, sortOptions: setSort(payload.field, payload.order) };
    }

    default: {
      return state;
    }
  }
};

export const initialiseReducer = (props) => {
  const iState = (props && props.initialState) ? props.initialState : defaultState;
  const [state, dispatch] = useReducer(reducer, iState);
  return { state, dispatch };
}