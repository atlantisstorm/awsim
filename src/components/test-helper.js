import React from 'react';
//import { render } from "@testing-library/react";
import Context from "../store/context";
import { initialiseReducer } from "../store/reducer";

export const RenderWithProvider = (props) => {
  const { children, testState, testDispatch }  = props;
  // Note: we pass any supplied testState to initialiseReducer rather than switching it like
  // we do for dispatch because of scoping.  Basically each rerender after a click/select
  // update event would mean we refresh again with the original data.
  const { state, dispatch: iDispatch } = initialiseReducer({ initialState: testState });
  const dispatch = testDispatch ? testDispatch : iDispatch;
  return (
    <Context.Provider value={{ state, dispatch }}>
      { children }
    </Context.Provider> 
  )
}