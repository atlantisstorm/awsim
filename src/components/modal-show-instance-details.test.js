import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import ModalShowInstanceDetails from './modal-show-instance-details';
import defaultInstanceRawData from '../fixtures/default-instance-raw-data';
import { defaultState } from '../store/model';
import { MODAL_SET_DISPLAY } from '../store/types';
import { RenderWithProvider } from './test-helper';

describe('ModalShowInstanceDetails component', () => {
  const testState = {
    modalSettings: {
      ...defaultState.modalSettings,
      instance: {
        ...defaultInstanceRawData,
        Tags: [
          { Key: "Tag1", Value: "Value1" },
          { Key: "Tag2", Value: "Value2" },
          { Key: "Tag3", Value: "Value3" }
        ],
        IamInstanceProfile: { Id: "Id value", Arm: "Arm value" }
      }
    }
  };

  it('should match snapshot', () => {
    const { asFragment } = render(
      <RenderWithProvider testState={testState}>
        <ModalShowInstanceDetails />
      </RenderWithProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call dispatch to make modal display false on click event', () => {
    const dispatch = jest.fn();
    const { getByTestId } = render(
      <RenderWithProvider testDispatch={dispatch} testState={testState}>
        <ModalShowInstanceDetails />
      </RenderWithProvider>
    );

    const modalClose = getByTestId('modal-close');
    fireEvent.click(modalClose);
    const payload = { "display": false };
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ type: MODAL_SET_DISPLAY, payload });
  });

  it('should contain expected section headers, field title and values', () => {
    const { container, asFragment } = render(
      <RenderWithProvider testState={testState}>
        <ModalShowInstanceDetails />
      </RenderWithProvider>
    );

    // Check section headers are in place.
    expect(container.innerHTML).toContain('<td colspan="2" class="header">General</td>');
    expect(container.innerHTML).toContain('<tr><td class="title">Architecture</td><td class="data">x86_64</td></tr>');

    expect(container.innerHTML).toContain('<td colspan="2" class="header">Tags</td>');
    expect(container.innerHTML).toContain('<tr><td class="title">Tag1</td><td class="data">Value1</td></tr>');

    expect(container.innerHTML).toContain('<td colspan="2" class="header">Iam Instance Profile</td>');
    expect(container.innerHTML).toContain('<tr><td class="title">Id</td><td class="data">Id value</td></tr>');

  });
});