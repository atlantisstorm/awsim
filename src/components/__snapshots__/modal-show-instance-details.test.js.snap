// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalShowInstanceDetails component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="context-menu"
    data-testid="modal-close"
    style="top: 5px; left: 5px;"
  >
    <div
      id="instance-tags"
    >
      <table>
        <tbody>
          <tr>
            <td
              class="header"
              colspan="2"
            >
              General
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Architecture
            </td>
            <td
              class="data"
            >
              x86_64
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Hypervisor
            </td>
            <td
              class="data"
            >
              xen
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Name
            </td>
            <td
              class="data"
            />
          </tr>
          <tr>
            <td
              class="title"
            >
              PrivateDnsName
            </td>
            <td
              class="data"
            >
              ip-192-168-34-117.eu-west-1.compute.internal
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              PublicDnsName
            </td>
            <td
              class="data"
            >
              ec2-54-42-117-131.eu-west-1.compute.amazonaws.com
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              RootDeviceType
            </td>
            <td
              class="data"
            >
              ebs
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              RootDeviceName
            </td>
            <td
              class="data"
            >
              /dev/sda1
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              VirtualizationType
            </td>
            <td
              class="data"
            >
              hvm
            </td>
          </tr>
          <tr>
            <td
              class="header"
              colspan="2"
            >
              Tags
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Tag1
            </td>
            <td
              class="data"
            >
              Value1
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Tag2
            </td>
            <td
              class="data"
            >
              Value2
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Tag3
            </td>
            <td
              class="data"
            >
              Value3
            </td>
          </tr>
          <tr>
            <td
              class="header"
              colspan="2"
            >
              Iam Instance Profile
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Id
            </td>
            <td
              class="data"
            >
              Id value
            </td>
          </tr>
          <tr>
            <td
              class="title"
            >
              Arn
            </td>
            <td
              class="data"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
